fmriprep/interfaces/utils.py

```diff
@@ -10,11 +10,75 @@
 import os
 import numpy as np
 import os.path as op
+import nibabel as nb
+from nipype.algorithms.confounds import is_outlier
+from nipype.interfaces.afni import Volreg
 from nipype.interfaces.base import (traits, isdefined, TraitedSpec, BaseInterface,
                                     BaseInterfaceInputSpec, File, InputMultiPath,
                                     OutputMultiPath)
 from nipype.interfaces import fsl
 
+from fmriprep.interfaces.bids import SimpleInterface
+
+
+class EstimateReferenceImageInputSpec(BaseInterfaceInputSpec):
+    in_file = File(exists=True, mandatory=True, desc="4D EPI file")
+
+
+class EstimateReferenceImageOutputSpec(TraitedSpec):
+    ref_image = File(exists=True, desc="3D reference image")
+    n_volumes_to_discard = traits.Int(desc="Number of detected non-steady "
+                                           "state volumes in the beginning of "
+                                           "the input file")
+
+
+class EstimateReferenceImage(SimpleInterface):
+    """
+    Given an 4D EPI file estimate an optimal reference image that could be later
+    used for motion estimation and coregistration purposes. If detected uses
+    T1 saturated volumes (non-steady state) otherwise a median of
+    of a subset of motion corrected volumes is used.
+    """
+    input_spec = EstimateReferenceImageInputSpec
+    output_spec = EstimateReferenceImageOutputSpec
+
+    def _run_interface(self, runtime):
+        in_nii = nb.load(self.inputs.in_file)
+        global_signal = in_nii.get_data()[:, :, :, :50].mean(axis=0).mean(
+            axis=0).mean(axis=0)
+
+        n_volumes_to_discard = is_outlier(global_signal)
+
+        out_ref_fname = os.path.abspath("ref_image.nii.gz")
+
+        if n_volumes_to_discard == 0:
+            if in_nii.shape[-1] > 40:
+                slice = in_nii.get_data()[:, :, :, 20:40]
+                slice_fname = os.path.abspath("slice.nii.gz")
+                nb.Nifti1Image(slice, in_nii.affine,
+                               in_nii.header).to_filename(slice_fname)
+            else:
+                slice_fname = self.inputs.in_file
+
+            res = Volreg(in_file=slice_fname, args='-Fourier -twopass', zpad=4,
```

Comment:
Why not mcflirt here again? (just a question, I assume 3dvolreg works better here)